Improve how libraries are located and various cleanup

tarsius opened the pull request:
Hi Tom, it was great talking to you at FOSDEM.  I have finally gotten around to playing with this ffi and I'm liking it a lot but I immediately ran into issues.

I think it has a lot of potential but also needs some work to fulfill its potential.  Even though this has been sitting around mostly unused for five years now, I hope that you are soon going to put in some more work to make it more usable.  It seems to me that the hard parts are done but that the polish is lacking.

I would like to help with that effort starting with this pull-request.  It comes with assorted cleanup, but the main improvements concern how libraries are located.  That is done in the last three commits.

It's probably not a good idea in the long run to re-purpose `load-path` for C libraries as well but at least it is a step up from forcing users to place C libraries inside this repository or to bind `default-directory` around code that might load a library. Maybe something like `ffi-library-path` should be added.

The module support in Emacs has the same issue more or less; here too it is not really specified how one should located a library.  As far as I can tell something like `module-path` exists.

I would like to see this added to Melpa rather sooner than later and I can help with that (I am a deputy maintainer).  But you would also have to put in more work of course.  I think it is worth it though.

This ffi could help with adaption of the dynamic module functionality a lot, even if it is only used for prototyping.  But that of course assumes that it is easy to get started.  We are not there yet but with more documentation and fewer sharp edges this could be quite successful.

We would also have to think about how a `ffi.el` that has been installed from Melpa would build `ffi-module.c` and how to communicate to the user that they have to install `libffi`.  There are a few modules in Melpa, but I don't think a common pattern has emerged yet how this sort of thing should be handled.  Eventually packages will appear on Melpa that use *this* package to interface with C, and we should avoid the same situation where everyone has to reinvent the wheel and should instead come up with best practices beforehand and document them.

So yeah, it will be a bit of work, but still: What do you think?

Ps: Please start by adding a LICENSE file and adding the permission statement to `ffi.el`.

nnicandro committed:
[WIP] Perform work in macro calling environment

In `define-ffi-function' and `ffi--struct-union-helper' the work to
define ffi functions (using `ffi--prep-cif' and `ffi--dlsym') and types
was done in the macro environment. This works fine during runtime, but
fails for complied code since the objects would be created at compile
time instead of at runtime. By moving the work to the macro's calling
environment we avoid this problem.

Co-authored-by: Jonas Bernoulli <jonas@bernoul.li>

tarsius committed:
Cleanup whitespace

tarsius committed:
Add library header and update documentation

tarsius committed:
make: Clean more generated files

tarsius committed:
make: Include config.mk if present

tarsius committed:
make: Extend the scope of the "all" target

tarsius committed:
emacs_module_init: Return early on subsequent invocations

Initializing the module a second time results in a segfault.

An attempt to initialize it a second time would occur when we evaluate
"ffi.el" a second time (e.g., if we `eval-buffer' after making changes
to some lisp code during development).

tarsius committed:
Don't use let* when let is enough

tarsius commented:
TL;DR With some extra effort, you might very well be responsible for a second game changer after `package.el`. :grinning:

tromey commented on ffi-module.c:
```diff
@@ -989,6 +988,18 @@ init_type_alias (const char *name, bool is_unsigned, int size)
+  env->funcall (env, Qprovide, 1, args);
+}
+
+static bool initialized = false;
```
I think you probably need `<stdbool.h>` to use `bool` and `false`; but it's just as clear (IMO) with `int` and less finicky.

tarsius replied:
I've done the former. I am not a C programmer and advice on the internets is unclear on what  else to do. I would have used `initialized == 1` but I read that that is a big no no. Please feel free to change this to your preferred style.

tromey commented on ffi.el:
```diff
@@ -1,21 +1,40 @@
+
+;; Copyright (C) 2015-2017 Tom Tromey
+
+;; Author: Tom Tromey <tromey@redhat.com>
```
I haven't been at this address in some years now.

tromey commented on ffi.el:
```diff
@@ -1,21 +1,40 @@
+;; Author: Tom Tromey <tromey@redhat.com>
+;; Package-Requires: ((emacs "25.1"))
+
+;; TODO Put permission statement here and also add LICENSE file.
```
I saw your note and I will do this.

tromey commented on ffi.el:
```diff
@@ -1,21 +1,40 @@
        (or ,library
-	   (setq ,library (ffi--dlopen ,name))))))
+	   (setq ,library
+		 (ffi--dlopen (or ,filename (ffi--locate-module ,name))))))))
```
I saw your note about `ffi-library-path` and I also wonder if that would be better.  Also it's probably important to allow loading of libraries from the system library paths -- that is, let the user pass in a relative path and hope that `dlopen` finds it.

tromey reviewed (commented):
Thanks for the patch.  I had a couple of nits but I think this is basically ready.  It's usually better for me if you send smaller, independent patches, since the simpler ones (whitespace cleanup or whatever) I tend to land immediately.  I'll reply to some other comments in the PR.

tromey commented:
>  I would also have to think about how a ffi.el that has been installed from Melpa would build ffi-module.c and how to communicate to the user that they have to install libffi

I wonder about this as well.

It would be possible to compile the FFI module's C code on first use.  However, as you point out, this is more difficult because libffi is also needed.  Really I was hoping to get this into the core, but I tend to doubt that will happen.  Maybe some elisp could look for the usual package managers and suggest something.  Or maybe most systems have libffi already anyway.

tarsius commented:
> It's usually better for me if you send smaller, independent patches, since the simpler ones (whitespace cleanup or whatever) I tend to land immediately.

I understand where you are coming from but I feel that this would just cause more work for *both of us*. I usually try to do it anyway, because it is understood as the right thing to do and I don't enjoy explaining why I didn't do it. But in this case I did not clearly see the line that splits the trivial changes from the stuff that might need more thought.

If you want me to do a second pull-request with just the trivial changes, then tell me where those end in your opinion.

Do you use the Github web interface to merge pull-requests? If so then yes I can see how in one case you can just click a button and in the other its not possible to merge at all until everything is done. But in that case I would recommend you stop doing that. Github is goddamn awful interface for doing merges or doing any kind of commits. See [this](https://github.com/magit/magit/wiki/Don't-commit-using-Github) if you want to read a rant about how bad it is.

And of course I am biased as the author of Magit and Forge, but it is still worth pointing out how superior that interface is. I can create a properly configured branch for a pull-request some asks me to merge by pressing `b y RET`. I can trivially merge just some of these commits. I can make trivial changes to someone else's commits myself and push the result to the original pull-request branch.

I recommend you give it a try, though maybe not right now. Instead, what exactly would you like me to do to make it easier for you to deal with this pull-request?

tarsius commented:
It would probably be best if a module that relies on this could be loaded like a module that does not. In other words it should be loadable using `(require 'MODULE)`, which does not only implies that it should be search for on the `load-path` but also that loading it like so should work even if nothing like `(module-load (ffi--locate-module "ffi-module"))` has run yet, i.e. the the ffi-enabled module would have to load the ffi-enabling module itself if necessary. Would that be possible?

suhail-singh commented:
@tarsius now that Tom has licensed the code under GPLv3 (after some gentle
nagging on IRC), I have taken the liberty of forking the project to
<https://github.com/emacs-ffi/emacs-ffi>.  I am in the process of assimilating
outstanding changes into my fork (see <https://github.com/emacs-ffi/emacs-ffi/issues/1>)
at the moment.

Would you be comfortable making a (possibly revised) PR against the
<https://github.com/emacs-ffi/emacs-ffi> fork?

tarsius committed:
test.el: Fix compiler error and warnings

- On Emacs 31 `lexical-binding' has to be set in every file.
- The `struct-type' variables were unused.

tarsius committed:
[KLUDGE] module: Include config.h

Without this we get many errors like this:

In file included from /.../emacs-module.h:29,
                 from ffi-module.c:16:
/.../lib/stddef.h:128:3: error: #error "Please include config.h first."
  128 |  #error "Please include config.h first."
      |   ^~~~~

All I know about this is that the compiler told me to do it and after
I did so, it stopped complaining.

tarsius commented:
Thanks for adding/getting added the license!

I already noticed that there was some movement and re-opened the pull-requests to ensure you (@suhail-singh) would see them. I'm going to open a pr in your repository soon.

tarsius committed:
define-ffi-{library,function}: Use same name for variable and function

Spamming the global namespace with symbols like `G123' is not okay.

Also define the variables at run-time, using `defvar'.

tarsius committed:
Declare indent rule for more define-ffi-*

Looking at the tests, we can see that that was the intention.

suhail-singh committed:
Merge remote-tracking branch 'tarsius/harmless' into dev

* .gitignore: Ignore config.mk.
* Makefile: Optionally allow for inclusion of config.mk.
(EMACS_BUILDDIR): Only optionally set EMACS_BUILDDIR.
(all): Include test-module during all target.
* ffi.el: Add Copyright, Package-Requires, Commentary headers.

tarsius committed:
lisp: Consistently use spaces for indentation

Also explicitly set `indent-tabs-mode' to nil, to avoid inconsistency
going forward.

tarsius committed:
test.el: Require ert

This isn't strictly necessary because the used forms are autoloaded,
but it seems cleaner to be explicit about it.  Replying on autoloads
seems more appropriate if the tests are defined in a proper library,
not in a file dedicated to tests.

tarsius committed:
Load our module at compile-time as well

We use functions from the module in "ffi.el",
so the byte-compiler needs to know about them.

tarsius committed:
Locate modules on the load-path

We have to do this for the libraries used by "ffi libraries"
as well as for our own "ffi-module.so".

Without doing it for the latter, we used to get:

  $ emacs -Q --batch -L . -f batch-byte-compile ffi.el

  In toplevel form:
  ffi.el:31:2: Error: Module could not be opened: \
  "ffi-module.so", "ffi-module.so: cannot open shared object file: \
  No such file or directory"

Note that in this case we are actually in the directory that contains
the module but never-the-less it doesn't work with a relative name.

tarsius committed:
ffi--struct-union-helper: Cosmetics

tarsius committed:
ffi--lay-out-struct: Cosmetics

tarsius committed:
define-ffi-function: Cosmetics

tarsius committed:
define-ffi-function: Name arguments consistently

Name them `arg<N>', where <N> is 1 initially and then is increased by 1
for every argument.  Previously `G<N>' but while <N> was also increased
for every argument, it started out at a random number, so the arguments
would get different names every time this is evaluated/compiled.

tarsius committed:
[WIP] Improve Makefile

tarsius committed:
[WIP] CI

tarsius committed:
[wip]
